// generator-jhipster/dist/generators/java/generators/domain/entity-files.js
import { javaMainPackageTemplatesBlock, javaTestPackageTemplatesBlock } from '../../support/index.js';
export const entityServerFiles = {
    model: [
        javaMainPackageTemplatesBlock({
            condition: ctx => ctx.entityDomainLayer,
            templates: ['_entityPackage_/domain/_persistClass_.java.jhi'],
        }),
    ],
    modelTestFiles: [
        javaTestPackageTemplatesBlock({
            condition: ctx => ctx.entityDomainLayer,
            templates: [
                '_entityPackage_/domain/_persistClass_Asserts.java',
                '_entityPackage_/domain/_persistClass_Test.java',
                '_entityPackage_/domain/_persistClass_TestSamples.java',
            ],
        }),
    ],
    server: [
        javaMainPackageTemplatesBlock({
            condition: ctx => ctx.useJakartaValidation && ctx.entityDomainLayer,
            templates: ['_entityPackage_/domain/_persistClass_.java.jhi.jakarta_validation'],
        }),
        javaMainPackageTemplatesBlock({
            condition: ctx => ctx.useJacksonIdentityInfo && ctx.entityDomainLayer,
            templates: ['_entityPackage_/domain/_persistClass_.java.jhi.jackson_identity_info'],
        }),
    ],
};
export const enumFiles = {
    enumFiles: [
        javaMainPackageTemplatesBlock({
            renameTo: (data, filepath) => filepath.replace('_enumName_', data.enumName),
            templates: ['_entityPackage_/domain/enumeration/_enumName_.java'],
        }),
    ],
};
