// generator-jhipster/dist/generators/base/command.js
const command = {
    options: {
        useVersionPlaceholders: {
            description: 'replace mutable versions with placeholders',
            type: Boolean,
            env: 'VERSION_PLACEHOLDERS',
            scope: 'generator',
            hide: true,
        },
        skipChecks: {
            description: 'Check the status of the required tools',
            type: Boolean,
            scope: 'generator',
        },
        experimental: {
            description: 'Enable experimental features. Please note that these features may be unstable and may undergo breaking changes at any time',
            type: Boolean,
            scope: 'generator',
        },
        disableBlueprints: {
            description: 'Disable blueprints support',
            type: Boolean,
            scope: 'generator',
        },
        debugEnabled: {
            name: 'debug',
            description: 'Enable debugger',
            alias: 'd',
            type: Boolean,
            scope: 'generator',
        },
        reproducible: {
            description: 'Try to reproduce changelog',
            type: Boolean,
            scope: 'control',
        },
        skipPrompts: {
            description: 'Skip prompts',
            type: Boolean,
            scope: 'generator',
        },
        ignoreNeedlesError: {
            description: 'Ignore needles failures',
            type: Boolean,
            hide: true,
            scope: 'generator',
        },
    },
};
export default command;
