// generator-jhipster/dist/generators/javascript/generators/prettier/command.js
import gitCommand from '../../../git/command.js';
const { monorepository } = gitCommand.configs;
const command = {
    configs: {
        fromInit: {
            description: 'Generate prettier config using init defaults',
            cli: {
                type: Boolean,
                hide: true,
            },
            scope: 'generator',
        },
        prettierConfigFile: {
            description: 'Prettier configuration file',
            cli: {
                type: Boolean,
                hide: true,
            },
            configure: (gen) => {
                gen.prettierConfigFile = gen.fromInit || gen.monorepositoryRoot ? '.prettierrc.yml' : '.prettierrc';
            },
            scope: 'generator',
        },
        prettierTabWidth: {
            description: 'Default tab width for prettier',
            cli: {
                type: Number,
            },
            default: 2,
            scope: 'storage',
        },
        monorepository,
        monorepositoryRoot: {
            cli: {
                type: Boolean,
                hide: true,
            },
            scope: 'generator',
        },
    },
    import: [],
};
export default command;
