// generator-jhipster/dist/generators/languages/command.js
const command = {
    arguments: {
        languages: {
            description: 'Languages to generate',
            type: Array,
            required: false,
        },
    },
    configs: {
        languages: {
            cli: {
                type: Array,
                hide: true,
            },
            scope: 'storage',
        },
        enableTranslation: {
            cli: {
                description: 'Enable translation',
                type: Boolean,
            },
            scope: 'storage',
        },
        language: {
            cli: {
                alias: 'l',
                description: 'Language to be added to application (existing languages are not removed)',
                type: Array,
            },
            scope: 'none',
        },
        nativeLanguage: {
            cli: {
                alias: 'n',
                description: 'Set application native language',
                type: String,
                required: false,
            },
            scope: 'storage',
        },
        regenerateLanguages: {
            cli: {
                description: 'Regenerate languages',
                type: Boolean,
            },
            scope: 'generator',
        },
    },
};
export default command;
