// generator-jhipster/dist/generators/generate-blueprint/command.js
import { GENERATOR_INIT } from '../generator-list.js';
import { ADDITIONAL_SUB_GENERATORS, ALL_GENERATORS, ALL_PRIORITIES, CLI_OPTION, DYNAMIC, GENERATE_SNAPSHOTS, JS, LINK_JHIPSTER_DEPENDENCY, LOCAL_BLUEPRINT_OPTION, SUB_GENERATORS, } from './constants.js';
const command = {
    configs: {
        githubRepository: {
            cli: {
                description: 'Github Repository',
                type: String,
            },
            scope: 'storage',
        },
        githubWorkflows: {
            cli: {
                description: 'Generate github workflows',
                type: Boolean,
            },
            scope: 'storage',
        },
    },
    options: {
        [GENERATE_SNAPSHOTS]: {
            description: 'Generate test snapshots',
            type: Boolean,
        },
        [LINK_JHIPSTER_DEPENDENCY]: {
            description: 'Link JHipster dependency for testing',
            type: Boolean,
            hide: true,
        },
        [SUB_GENERATORS]: {
            description: 'Sub generators to generate',
            type: Array,
            scope: 'storage',
        },
        [ADDITIONAL_SUB_GENERATORS]: {
            description: 'Comma separated additional sub generators to generate',
            type: String,
            scope: 'storage',
        },
        [DYNAMIC]: {
            description: 'Generate dynamic generators (advanced)',
            type: Boolean,
            scope: 'storage',
        },
        [JS]: {
            description: 'Use js extension',
            type: Boolean,
            scope: 'storage',
        },
        [LOCAL_BLUEPRINT_OPTION]: {
            description: 'Generate a local blueprint',
            type: Boolean,
            scope: 'storage',
        },
        [CLI_OPTION]: {
            description: 'Generate a cli for the blueprint',
            type: Boolean,
            scope: 'storage',
        },
        [ALL_GENERATORS]: {
            description: 'Generate every sub generator',
            type: Boolean,
            scope: 'generator',
        },
        [ALL_PRIORITIES]: {
            description: 'Generate every priority',
            type: Boolean,
        },
    },
    import: [GENERATOR_INIT],
};
export default command;
