// generator-jhipster/dist/generators/base-application/support/task-type-inference.js
export function asWriteFilesSection(section) {
    return section;
}
export function asWriteFilesBlock(section) {
    return section;
}
export function asInitializingTask(task) {
    return task;
}
export function asPromptingTask(task) {
    return task;
}
export function asConfiguringTask(task) {
    return task;
}
export function asComposingTask(task) {
    return task;
}
export function asLoadingTask(task) {
    return task;
}
export function asPreparingTask(task) {
    return task;
}
export function asPostPreparingTask(task) {
    return task;
}
export function asPreparingEachEntityTask(task) {
    return task;
}
export function asPreparingEachEntityFieldTask(task) {
    return task;
}
export function asPreparingEachEntityRelationshipTask(task) {
    return task;
}
export function asPostPreparingEachEntityTask(task) {
    return task;
}
export function asDefaultTask(task) {
    return task;
}
export function asWritingTask(task) {
    return task;
}
export function asWritingEntitiesTask(task) {
    return task;
}
export function asPostWritingTask(task) {
    return task;
}
export function asPostWritingEntitiesTask(task) {
    return task;
}
export function asInstallTask(task) {
    return task;
}
export function asEndTask(task) {
    return task;
}
