// generator-jhipster/dist/generators/liquibase/files.js
import { SERVER_MAIN_RES_DIR, SERVER_MAIN_SRC_DIR } from '../generator-constants.js';
import { moveToJavaPackageSrcDir } from '../server/support/index.js';
export const liquibaseFiles = {
    liquibase: [
        {
            condition: ctx => ctx.backendTypeSpringBoot,
            path: `${SERVER_MAIN_SRC_DIR}_package_/`,
            renameTo: moveToJavaPackageSrcDir,
            templates: ['config/LiquibaseConfiguration.java'],
        },
    ],
    gradle: [
        {
            condition: ctx => ctx.buildToolGradle,
            templates: ['gradle/liquibase.gradle'],
        },
    ],
    serverResource: [
        {
            path: SERVER_MAIN_RES_DIR,
            templates: [
                {
                    override: ctx => !ctx.incrementalChangelog || ctx.recreateInitialChangelog,
                    file: data => `config/liquibase/changelog/initial_schema_${data.databaseType}.xml`,
                    renameTo: () => 'config/liquibase/changelog/00000000000000_initial_schema.xml',
                },
                {
                    override: ctx => !ctx.incrementalChangelog || ctx.recreateInitialChangelog,
                    file: 'config/liquibase/master.xml',
                },
            ],
        },
        {
            condition: generator => Boolean(generator.generateUserManagement),
            path: SERVER_MAIN_RES_DIR,
            templates: ['config/liquibase/data/user.csv', 'config/liquibase/data/user_authority.csv'],
        },
        {
            condition: generator => Boolean(generator.generateBuiltInAuthorityEntity),
            path: SERVER_MAIN_RES_DIR,
            templates: ['config/liquibase/data/authority.csv'],
        },
    ],
};
